vsphere/resource_vsphere_vnic.go

```diff
@@ -0,0 +1,517 @@
+	"github.com/vmware/govmomi"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/types"
+	"log"
+	"strconv"
+	"strings"
+)
+
+func resourceVsphereNic() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceVsphereNicCreate,
+		Read:   resourceVsphereNicRead,
+		Update: resourceVsphereNicUpdate,
+		Delete: resourceVsphereNicDelete,
+		Importer: &schema.ResourceImporter{
+			State: schema.ImportStatePassthrough,
+		},
+		Schema: vNicSchema(),
+	}
+}
+
+func vNicSchema() map[string]*schema.Schema {
+	base := BaseVMKernelSchema()
+	base["host"] = &schema.Schema{
+		Type:        schema.TypeString,
+		Required:    true,
+		Description: "ESX host the interface belongs to",
+		ForceNew:    true,
+	}
+
+	return base
+}
+
+func resourceVsphereNicRead(d *schema.ResourceData, meta interface{}) error {
+	ctx := context.TODO()
+	client := meta.(*VSphereClient).vimClient
+	tfNicID := d.Id()
+
+	toks := strings.Split(tfNicID, "_")
+	hostId := toks[0]
+	nicId := toks[1]
+
+	vnic, err := getVnicFromHost(ctx, client, hostId, nicId)
+	if err != nil {
+		log.Printf("[DEBUG] Nic (%s) not found. Probably deleted.", nicId)
+		d.SetId("")
+		return nil
+	}
+
+	_ = d.Set("portgroup", vnic.Portgroup)
+	if vnic.Spec.DistributedVirtualPort != nil {
+		_ = d.Set("distributed_switch_port", vnic.Spec.DistributedVirtualPort.SwitchUuid)
+		_ = d.Set("distributed_port_group", vnic.Spec.DistributedVirtualPort.PortgroupKey)
+	}
+	_ = d.Set("mtu", vnic.Spec.Mtu)
+	_ = d.Set("mac", vnic.Spec.Mac)
+	//_ = d.Set("netstack", vnic.Spec.NetStackInstanceKey)
+
+	// Do we have any ipv4 config ?
+	if _, ok := d.GetOk("ipv4"); ok {
+		// if DHCP is true then we should ignore whatever addresses are set here.
+		ipv4dict := make(map[string]interface{})
+		ipv4dict["dhcp"] = vnic.Spec.Ip.Dhcp
+		if !vnic.Spec.Ip.Dhcp {
+			ipv4dict["ip"] = vnic.Spec.Ip.IpAddress
+			ipv4dict["netmask"] = vnic.Spec.Ip.SubnetMask
+			if vnic.Spec.IpRouteSpec != nil {
+				ipv4dict["gw"] = vnic.Spec.IpRouteSpec.IpRouteConfig.GetHostIpRouteConfig().DefaultGateway
+			}
+		}
+		_ = d.Set("ipv4", []map[string]interface{}{ipv4dict})
+	}
+
+	// Do we have any ipv6 config ?
+	if _, ok := d.GetOk("ipv6"); ok {
+		ipv6dict := map[string]interface{}{
+			"dhcp":       *vnic.Spec.Ip.IpV6Config.DhcpV6Enabled,
+			"autoconfig": *vnic.Spec.Ip.IpV6Config.AutoConfigurationEnabled,
+		}
+		// First we need to filter out addresses that were configured via dhcp or autoconfig
+		// or link local or any other mechanism
+		addrList := make([]string, 0)
+		for _, addr := range vnic.Spec.Ip.IpV6Config.IpV6Address {
+			if addr.Origin == "manual" {
+				addrList = append(addrList, fmt.Sprintf("%s/%d", addr.IpAddress, addr.PrefixLength))
+			}
+		}
+		ipv6dict["addresses"] = addrList
+		if vnic.Spec.IpRouteSpec != nil {
+			ipv6dict["gw"] = vnic.Spec.IpRouteSpec.IpRouteConfig.GetHostIpRouteConfig().IpV6DefaultGateway
+		} else if _, ok := d.GetOk("ipv6.0.gw"); ok {
```

Comment:
Why is `gw` only set if it exists in the config? If it isn't in the config, it can still be set to "", right?

Reply:
If we don't set one in the config then there shouldn't be one set on the host either. Otherwise someone could set the gw via the vsphere UI and cause a configuration drift that we'd be unable to remediate.

Reply:
Maybe I'm misreading, but this saying that if `ipv6.0.gw` is not set in the state or config and the `spec` is `nil`, then set `ipv6.0.gw` to `""`, right? Wouldn't we want it to set `ipv6.0.gw` to `""` any time vsphere says `IpRouteSpec` is empty?